### ISSUE TITLE
Feature request: Only perform a checkout when it is required for paths-filter to operate.

### ISSUE THREAD

**Issue by dcopso** [dorny/paths-filter#198]
Per https://github.com/dorny/paths-filter#supported-workflows, only the "Feature branches" and "Master, Release, or other long-lived branches" scenarios require the code to be checked out. 

On some projects, a checkout can take a long time (on mine, sometimes 60s+). It would be nice if the conditions that a checkout was required could be expressed as a GitHub Actions expression for use in an `if:` statement on the checkout step to conditionally control whether the expensive checkout occurs before paths-filter.

Are all the conditions for which git commands must be used testable via GHA expressions?

Thanks. Love this action, it's saved us so much time!

**Comment by zymotik** [comment]
Would this action suit you better @dcopso? [tj-actions/changed-files](https://github.com/tj-actions/changed-files)

**Comment by saschabratton** [comment]
Yeah, I think you can actually get this today with a single `if:` on the checkout step:

```yml
- uses: actions/checkout@v4
  if: ${{ !startsWith(github.event_name, 'pull_request') }}
```

The one case where paths-filter _doesn't_ need a checkout is when it can pull the list of changed files straight from the GitHub API instead of using git... and that only happens on the PR-family events (`pull_request`, `pull_request_target`, `pull_request_review`, `pull_request_review_comment`) when a token is available (which it is by default via `github.token`). Everything else (push, workflow_dispatch, schedule, etc.) falls back to git diff, so it needs the code checked out. Since all those PR events start with `pull_request`, the `startsWith(...)` covers them in one go.

Couple caveats to watch out for...
- if you explicitly set `token: ''` to force git-based diffing on PRs, then you _do_ still need the checkout on PR events, so this wouldn't apply.
- same if you use `base: HEAD` to detect uncommitted changes... that path always needs a checkout.

But for the standard setup I think that one-liner gets you what you're after. lmk if it helps ☺️
